builtin/logical/database/backend.go

```diff
@@ -27,6 +29,29 @@ const (
 	minRootCredRollbackAge = 1 * time.Minute
 )
 
+// metrics collection
+var (
+	gaugeSync = sync.Mutex{}
+	gauges    = map[string]*int32{}
+	gaugeKey  = []string{"secrets", "database", "backend", "connections", "count"}
+)
+
+func createGauge(name string) {
+	gaugeSync.Lock()
+	defer gaugeSync.Unlock()
+	gauges[name] = new(int32)
+}
+
+func addToGauge(dbType, version string, uuid string, amount int32) {
+	labels := []metrics.Label{{"type", dbType}, {"version", version}, {"backend", uuid}}
+	gaugeName := fmt.Sprintf("%s-%s", dbType, version)
+	if _, ok := gauges[gaugeName]; !ok {
+		createGauge(gaugeName)
+	}
```

Comment:
None of the reads of `gauges` are inside a `gaugeSync` lock, but the writes are. It seems like all of the calls to `addToGauge` and `createGauge` are ultimately covered by the `connLock`, although that's per database mount and this is global. If you end up moving the gauge variables inside the database backend, then I think you could remove all the gauge locking? And if you keep it global, I think you can cover everything by turning `createGauge` into `createOrGetGauge` with an exclusive lock for the whole function body, and return a plain `*int32` value from the map.